aws/data_source_aws_iam_policy_document.go

```diff
@@ -112,6 +122,39 @@ func dataSourceAwsIamPolicyDocumentRead(d *schema.ResourceData, meta interface{}
 		}
 	}
 
+	// populate mergedDoc with any provided source_json_list
+	if sourceJSONList, hasSourceJSONList := d.GetOk("source_json_list"); hasSourceJSONList {
+
+		// generate sid map to assure there are no duplicates in source jsons
+		sidMap := make(map[string]struct{})
+		for _, stmt := range mergedDoc.Statements {
+			if stmt.Sid != "" {
+				sidMap[stmt.Sid] = struct{}{}
+			}
+		}
+
+		// merge sourceDocs in order specified
+		for sourceJSONIndex, sourceJSON := range sourceJSONList.([]interface{}) {
+			sourceDoc := &IAMPolicyDoc{}
+			if err := json.Unmarshal([]byte(sourceJSON.(string)), sourceDoc); err != nil {
+				return err
+			}
+
+			// assure all statements in sourceDoc are unique before merging
+			for stmtIndex, stmt := range sourceDoc.Statements {
+				if stmt.Sid != "" {
+					if _, sidExists := sidMap[stmt.Sid]; sidExists {
+						return fmt.Errorf("Found duplicate sid (%s) in source_json_list (item %d; statement %d). Either remove the sid or ensure the sid is unique across all statements.", stmt.Sid, sourceJSONIndex, stmtIndex)
```

Comment:
This is slightly different to the normal error messages. I wanted to include the `json` index and `statement` index for developer ease, but am open to discuss different ways to convey this information.